bwe/BandwidthEstimator.cpp

```diff
@@ -255,94 +355,90 @@ math::Matrix<double, 3> BandwidthEstimator::getCovariance() const
     return r;
 }
 
+// generate alternative current positions based on noise in model and process
 void BandwidthEstimator::generateSigmaPoints(const math::Matrix<double, 3>& state,
     const math::Matrix<double, 3, 3>& covP,
     const math::Matrix<double, 3>& processNoise,
     std::array<math::Matrix<double, 3>, SIGMA_POINTS>& sigmaPoints)
 {
+    assert(state(Bandwidth) > 10);
     static const auto seed = covP.I() * 0.0000001; // will make it positive definite
     const auto squareRoot = math::choleskyDecompositionLL(covP + seed);
-    sigmaPoints[0] = _state;
+    sigmaPoints[0] = state;
 
     int startIndex = 1;
+    const double maxBandwidthDeviation = std::max(0.0, state(Bandwidth) - 10);
     for (int c = 0; c < squareRoot.columns(); ++c)
     {
         auto sigmaOffset = _sigmaWeight * squareRoot.getColumn(c);
-        sigmaPoints[startIndex + 2 * c] = _state + sigmaOffset;
-        sigmaPoints[startIndex + 2 * c + 1] = _state - sigmaOffset;
+        sigmaOffset(QueuedBits) = math::clamp(sigmaOffset(QueuedBits), -state(QueuedBits), state(QueuedBits));
+        sigmaOffset(Bandwidth) = math::clamp(sigmaOffset(Bandwidth), -maxBandwidthDeviation, maxBandwidthDeviation);
+
+        sigmaPoints[startIndex + 2 * c] = state + sigmaOffset;
+        sigmaPoints[startIndex + 2 * c + 1] = state - sigmaOffset;
     }
 
     startIndex += 2 * squareRoot.columns();
-    for (int i = 0; i < _processNoise.rows(); ++i)
+    for (int i = 0; i < processNoise.rows(); ++i)
     {
         math::Matrix<double, 3> noise;
         noise(i) = processNoise(i) * _sigmaWeight;
-        sigmaPoints[startIndex + i * 2] = _state + noise;
-        // Offset the positive clock offset sigma point. Better catch up to clock drift.
-        // CO is easily pushed down by packets with short delay.
-        sigmaPoints[startIndex + i * 2 + 1] = _state - (i == 2 ? noise * 0.001 : noise);
```

Comment:
Tests show this is no longer needed. It was a problem caused by the non-linear transform. Clock offset accurately follows the clock drift as observed in real traces.